syntax = "proto3";

package bayesmech.vision;

option java_multiple_files = true;
option java_package = "com.bayesmech.vision";

import "primitives.proto";
import "spatial.proto";

// Message being sent from all Perceiver devices (i.e. phones with cameras and sensors).
message PerceiverDataFrame {
  // Metadata detailing which device this arframe comes from,
  // what frame number it is, and when it was captured.
  PerceiverFrameIdentifier frame_identifier = 1;

  // Camera pose in world coordinates
  Pose camera_pose = 2;

  // Image and Depth sensor data, identical in number of rows x cols.
  ImageFrame rgb_frame = 3;
  DepthFrame depth_frame = 4;

  // IMU data, accelerometer and gyroscope.
  ImuData imu_data = 5;

  // Details about the camera matrix, only populated in the first frame of communication
  CameraIntrinsics camera_intrinsics = 6;

  // ARCore-inferred spatial geometry (planes, point cloud).
  InferredGeometry inferred_geometry = 7;
}

message PerceiverFrameIdentifier {
  int64 timestamp_ns = 1;
  uint32 frame_number = 2;
  string device_id = 3;
}

message CameraIntrinsics {
  float fx = 1;
  float fy = 2;
  float cx = 3;
  float cy = 4;
  float image_width = 5;
  float image_height = 6;
  // Depth camera resolution (may differ from color camera resolution)
  float depth_width = 7;
  float depth_height = 8;
}

message ImageFrame {

  enum ImageFormat {
    UNKNOWN = 0;
    BITMAP_RGB = 1;
    BITMAP_RGBA = 2;
    YUV_420 = 3;
    JPEG = 4;
    GRAYSCALE = 5;
  }

  bytes data = 1;

  ImageFormat format = 2;
  uint32 quality = 5;  // Compression quality (0-100 for JPEG)
}

message DepthFrame {

  enum DepthFormat {
    DEPTH_FORMAT_UNKNOWN = 0;
    UINT16_MILLIMETERS = 1;
    FLOAT32_METERS = 2;
  }

  bytes data = 1;
  bytes confidence = 2;  // Confidence map (optional, same dimensions, 0-255 per pixel)

  DepthFormat format = 3;
}

message ImuData {
  Vector3 angular_velocity = 1;
  Vector3 linear_acceleration = 2;  // obtained from raw acceleration by ARCore primitive
  Vector3 gravity = 3;              // obtained from raw accelerometer by ARCore primitive
  Vector3 magnetic_field = 4;
}
